### PR TITLE
Create basic extension skeleton

### PR DESCRIPTION
**Pull request by senky**

### CODE REVIEW

**Comment by senky** [comment]
[Basic skeleton](https://trello.com/c/4jq1iOVw/4-basic-skeleton)

File: .travis.yml [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    - php: hhvm
+  fast_finish: true
+
+env:
+  global:
+    - EXTNAME="phpbb/admanagement"  # CHANGE name of the extension HERE
+    - SNIFF="1"            # Should we run code sniffer on your code?
+    - IMAGE_ICC="1"        # Should we run icc profile sniffer on your images?
+    - EPV="1"              # Should we run EPV (Extension Pre Validator) on your code?
+    - PHPBB_BRANCH="3.1.x"
```
Comment (VSEphpbb) [inline .travis.yml]: Lets update this for testing against phpBB 3.2 instead (all the other official extensions are set up for 3.2).

3.2 is the current main release of phpBB now, and 3.1.x is scheduled to end maintenence next month too.

File: acp/main_info.php [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+/**
+ * Advertisement management ACP module info.
+ */
+class main_info
+{
+	public function module()
+	{
+		return array(
+			'filename'	=> '\phpbb\admanagement\acp\main_module',
+			'title'		=> 'ACP_DEMO_TITLE',
```
Comment (VSEphpbb) [inline acp/main_info.php]: I would go ahead and update these language keys for ad management now.

File: acp/main_module.php [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+	public $page_title;
+	public $tpl_name;
+	public $u_action;
+
+	public function main($id, $mode)
+	{
+		global $config, $request, $template, $user;
+
+		$user->add_lang_ext('phpbb/admanagement', 'common');
+		$this->tpl_name = 'acp_demo_body';
```
Comment (VSEphpbb) [inline acp/main_module.php]: here too, I would rename this for our extension asap. Better to clean out what we can of the acme demo now than let it get lost.

File: composer.json [hunk 4]
```diff
@@ -0,0 +1,39 @@
+{
+    "name": "phpbb/admanagement",
+    "type": "phpbb-extension",
+    "description": "An extension for phpBB that allows administrators to add HTML/JS advertisement code to number of locations in the template.",
```
Comment (VSEphpbb) [inline composer.json]: I'd reword to say: ... allows administrators to manage and display advertisements on their forums.

File: composer.json [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+        },
+        {
+            "name": "Matt Friedman",
+            "homepage": "http://mattfriedman.me",
+            "role": "Extensions Development Team Lead"
+        },
+        {
+            "name": "Derky",
+            "homepage": "http://www.derky.nl",
+            "role": "Extensions Development Team"
```
Comment (VSEphpbb) [inline composer.json]: Just "Extensions Team" for Derky.

File: composer.json [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+        }
+    ],
+    "require": {
+        "php": ">=5.3.3",
+        "composer/installers": "~1.0"
+    },
+    "extra": {
+        "display-name": "Advertisement Management",
+        "soft-require": {
+            "phpbb/phpbb": ">=3.1.0,<3.3.0@dev"
```
Comment (VSEphpbb) [inline composer.json]: Just do `>=3.1.0` for now.

File: event/main_listener.php [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+ * @ignore
+ */
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+
+/**
+ * Advertisement management Event listener.
+ */
+class main_listener implements EventSubscriberInterface
+{
+	static public function getSubscribedEvents()
```
Comment (VSEphpbb) [inline event/main_listener.php]: Add a docblock with `{@inheritdoc}`

File: migrations/install_acp_module.php [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+
+	static public function depends_on()
+	{
+		return array('\phpbb\db\migration\data\v310\gold');
+	}
+
+	public function update_data()
+	{
+		return array(
+			array('config.add', array('acme_demo_goodbye', 0)),
```
Comment (VSEphpbb) [inline migrations/install_acp_module.php]: Lets plug in the correct lang key names here now. And dump whatever we don't need from this for now.

File: README.md [hunk 9]
```diff
@@ -1 +1,13 @@
-# ad-management
\ No newline at end of file
+# Advertisement Management
+
+This is the repository for the development of the Advertisement Management phpBB extension.
+
+[![Build Status](https://travis-ci.org/phpbb-extensions/ad-management.png)](https://travis-ci.org/phpbb-extensions/ad-management)
+[![Scrutinizer Code Quality](https://scrutinizer-ci.com/g/phpbb-extensions/ad-management/badges/quality-score.png?b=master)](https://scrutinizer-ci.com/g/phpbb-extensions/ad-management/?branch=master)
+[![Code Coverage](https://scrutinizer-ci.com/g/phpbb-extensions/ad-management/badges/coverage.png?b=master)](https://scrutinizer-ci.com/g/phpbb-extensions/ad-management/?branch=master)
```
Comment (VSEphpbb) [inline README.md]: if you want to test code coverage you have to look at Collapsoble Categories. Look at its travis folder and travis.yml file to see how coverage is setup.

File: migrations/install_user_schema.php [hunk 10]
```diff
@@ -0,0 +1,58 @@
+<?php
```
Comment (VSEphpbb) [inline migrations/install_user_schema.php]: I would eliminate this migration. It's not adding anything we need at the moment.

File: README.md (hunk 9 above)
Comment (senky) [inline README.md]: I updated all required files, but coverage is still not considered by scrutinizer. Do I need to add something to the repo, or you need to turn it on in scrutinizer?

File: .travis.yml [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+  - travis/setup-database.sh $DB $TRAVIS_PHP_VERSION $NOTESTS
+
+script:
+  - sh -c "if [ '$SNIFF' != '0' ]; then travis/ext-sniff.sh $DB $TRAVIS_PHP_VERSION $EXTNAME $NOTESTS; fi"
+  - sh -c "if [ '$IMAGE_ICC' != '0' ]; then travis/check-image-icc-profiles.sh $DB $TRAVIS_PHP_VERSION $NOTESTS; fi"
+  - sh -c "if [ '$EPV' != '0' ] && [ '$NOTESTS' = '1' ]; then phpBB/vendor/bin/EPV.php run --dir='phpBB/ext/$EXTNAME/'; fi"
+  - sh -c "if [ '$NOTESTS' != '1' ]; then ../$GITREPO/travis/run-tests.sh $DB $TRAVIS_PHP_VERSION $EXTNAME; fi"
+
+after_script:
+  - sh -c "if [ '$COVERAGE' != '0' ]; then ../$GITREPO/travis/send-coverage.sh $DB $TRAVIS_PHP_VERSION $GITREPO; fi"
```
Comment (VSEphpbb) [inline .travis.yml]: this file needs an empty newline feed at the end

File: acp/main_module.php [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+	public $u_action;
+
+	public function main($id, $mode)
+	{
+		global $config, $request, $template, $user;
+
+		$user->add_lang_ext('phpbb/admanagement', 'common');
+		$this->tpl_name = 'acp_admanagement';
+		$this->page_title = $user->lang('ACP_ADMANAGEMENT_TITLE');
+		add_form_key('acme/demo');
```
Comment (VSEphpbb) [inline acp/main_module.php]: change the form keys to phpbb/admanagement please

File: adm/style/acp_admanagement.html [hunk 13]
```diff
@@ -0,0 +1,24 @@
+<!-- INCLUDE overall_header.html -->
+
+<h1>{L_ACP_DEMO_TITLE}</h1>
+
+<form id="acp_board" method="post" action="{U_ACTION}">
```
Comment (VSEphpbb) [inline adm/style/acp_admanagement.html]: this ID should be something unique for the extension

File: migrations/install_acp_module.php [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+namespace phpbb\admanagement\migrations;
+
+class install_acp_module extends \phpbb\db\migration\migration
+{
+	public function effectively_installed()
+	{
+		$sql = 'SELECT module_id
+				FROM ' . $this->table_prefix . "modules
+				WHERE module_class = 'acp'
+					AND module_langname = 'ACP_ADMANAGEMENT_TITLE'";
```
Comment (VSEphpbb) [inline migrations/install_acp_module.php]: there seem to be extra tabs in these lines. should just be 1 tab per indent:

```php
$sql = 'SELECT module_id
	FROM ' . $this->table_prefix . "modules
	WHERE module_class = 'acp'
		AND module_langname = 'ACP_ADMANAGEMENT_TITLE'";
```

File: travis/prepare-coverage.sh [hunk 15]
```diff
@@ -0,0 +1,19 @@
...
+
+DB=$1
+TRAVIS_PHP_VERSION=$2
+EXTNAME=$3
+
+if [ "$TRAVIS_PHP_VERSION" == "5.5" -a "$DB" == "mysqli" ]
+then
+	sed -n '1h;1!H;${;g;s/<\/php>/<\/php>\n\t<filter>\n\t\t<whitelist>\n\t\t\t<directory>..\/<\/directory>\n\t\t\t<exclude>\n\t\t\t\t<directory>..\/tests\/<\/directory>\n\t\t\t\t<directory>..\/language\/<\/directory>\n\t\t\t\t<directory>..\/migrations\/<\/directory>\n\t\t\t<\/exclude>\n\t\t<\/whitelist>\n\t<\/filter>/g;p;}' phpBB/ext/$EXTNAME/travis/phpunit-mysqli-travis.xml &> phpBB/ext/$EXTNAME/travis/phpunit-mysqli-travis.xml.bak
+	cp phpBB/ext/$EXTNAME/travis/phpunit-mysqli-travis.xml.bak phpBB/ext/$EXTNAME/travis/phpunit-mysqli-travis.xml
+fi
```
Comment (VSEphpbb) [inline travis/prepare-coverage.sh]: newline needed at end of file

File: travis/prepare-epv.sh [hunk 16]
```diff
@@ -0,0 +1,22 @@
...
+
+EPV=$1
+NOTESTS=$2
+
+if [ "$EPV" == "1" -a "$NOTESTS" == "1" ]
+then
+	cd phpBB
+	composer require phpbb/epv:dev-master --dev --no-interaction
+	cd ../
+fi
```
Comment (VSEphpbb) [inline travis/prepare-epv.sh]: newline needed at end of file

File: travis/prepare-phpbb.sh [hunk 17]
```diff
@@ -0,0 +1,24 @@
...
+BRANCH=$2
+EXTPATH_TEMP=$3
+
+# Copy extension to a temp folder
+mkdir ../../tmp
+cp -R . ../../tmp
+cd ../../
+
+# Clone phpBB
+git clone --depth=1 "git://github.com/phpbb/phpbb.git" "phpBB3" --branch=$BRANCH
```
Comment (VSEphpbb) [inline travis/prepare-phpbb.sh]: newline needed at end of file

File: travis/run-tests.sh [hunk 18]
```diff
@@ -0,0 +1,20 @@
...
+DB=$1
+TRAVIS_PHP_VERSION=$2
+EXTNAME=$3
+
+if [ "$TRAVIS_PHP_VERSION" == "5.5" -a "$DB" == "mysqli" ]
+then
+	phpBB/vendor/bin/phpunit --configuration phpBB/ext/$EXTNAME/travis/phpunit-$DB-travis.xml --bootstrap ./tests/bootstrap.php --coverage-clover build/logs/clover.xml
+else
+	phpBB/vendor/bin/phpunit --configuration phpBB/ext/$EXTNAME/travis/phpunit-$DB-travis.xml --bootstrap ./tests/bootstrap.php
+fi
```
Comment (VSEphpbb) [inline travis/run-tests.sh]: newline needed at end of file

File: travis/send-coverage.sh [hunk 19]
```diff
@@ -0,0 +1,20 @@
...
+DB=$1
+TRAVIS_PHP_VERSION=$2
+GITREPO=$3
+
+if [ "$TRAVIS_PHP_VERSION" == "5.5" -a "$DB" == "mysqli" ]
+then
+    cd ../$GITREPO
+    wget https://scrutinizer-ci.com/ocular.phar
+    php ocular.phar code-coverage:upload --format=php-clover ../../phpBB3/build/logs/clover.xml
+fi
```
Comment (VSEphpbb) [inline travis/send-coverage.sh]: newline needed at end of file

**Comment by VSEphpbb** [comment]
Can you convert the templates to Twig please.

File: adm/style/acp_admanagement.html [hunk 20]
```diff
@@ -1,24 +1,24 @@
-<!-- INCLUDE overall_header.html -->
+{{ include('overall_header.html') }}
```
Comment (VSEphpbb) [inline adm/style/acp_admanagement.html]: Parenthesis are not needed

File: adm/style/acp_admanagement.html [hunk 21]
```diff
@@ -1,24 +1,24 @@
...
-		{S_FORM_TOKEN}
+		<input class="button1" type="submit" id="submit" name="submit" value="{{ lang('SUBMIT') }}" />&nbsp;
+		<input class="button2" type="submit" name="preview" value="{{ lang('PREVIEW') }}" />
+		{{ S_FORM_TOKEN }}
 	</fieldset>
 
 </form>
 
-<!-- INCLUDE overall_footer.html -->
+{{ include('overall_footer.html') }}
```
Comment (VSEphpbb) [inline adm/style/acp_admanagement.html]: Parenthesis not needed

**Comment by senky** [comment]
🎊
